cpp/src/arrow/array/concatenate.h

```diff
@@ -24,6 +24,23 @@
 #include "arrow/util/visibility.h"
 
 namespace arrow {
+namespace internal {
```

Comment:
why is this exported but internal? 🤔

Reply:
The cast suggestion is not stable yet. I think a better cast for strings would be strings views, but that can only be done when I push a PR that implements string->string-view casts (I have the code, but it's blocked by some half-float issues).

Reply:
Got this!